Log errors

kimlisa opened the pull request:
resolves https://github.com/gravitational/teleport.e/issues/208

#### Description
Provide better error handling

kimlisa committed:
Log errors

webvictim commented on lib/web/apiserver.go:
```diff
@@ -860,14 +850,10 @@ func (h *Handler) oidcLoginWeb(w http.ResponseWriter, r *http.Request, p httprou
-		}
-		http.Redirect(w, r, pathToError.String(), http.StatusFound)
-		return nil, nil
+		logger.WithError(err).Error("Error with creating auth request.")
```
To match other errors:

```suggestion
		logger.WithError(err).Error("Error creating auth request.")
```

kimlisa replied:
revised

kimlisa committed:
Address CR

kimlisa commented on lib/web/apiserver.go:
```diff
@@ -2328,3 +2315,39 @@ func makeTeleportClientConfig(ctx *SessionContext) (*client.Config, error) {
+	return nil, nil
+}
+
+func extractAuthConnQueryParamAndCSRFToken(r *http.Request) (clientRedirectURL, connectorID, csrfToken string, err error) {
```
revised

kimlisa commented on lib/web/apiserver.go:
```diff
@@ -2328,3 +2315,39 @@ func makeTeleportClientConfig(ctx *SessionContext) (*client.Config, error) {
+	return nil, nil
+}
+
+func redirectToErrorPage(w http.ResponseWriter, r *http.Request) (interface{}, error) {
```
revised, i created this helper function to help me do one liner return statements `return redirectToErrorPage`, instead of doing two liner like below since redirecting didn't stop code execution:
```
http.Redirect(w, r, loginFailedURL, http.Status.Found)
return nil, nil
```

alex-kovoy reviewed (approved):
with minor comment

kimlisa committed:
Change return type to pointer

r0mant commented on lib/web/apiserver.go:
```diff
@@ -2328,3 +2336,35 @@ func makeTeleportClientConfig(ctx *SessionContext) (*client.Config, error) {
+
+	connectorID := query.Get("connector_id")
+	if connectorID == "" {
+		return nil, trace.BadParameter("Missing connector_id query parameter.")
```
Do not capitalize and remove the dot. Same below.

kimlisa replied:
revised

r0mant commented on lib/web/apiserver.go:
```diff
@@ -64,6 +64,11 @@ import (
 )
 
+const (
+	loginFailedURL            = "/web/msg/error/login"
```
Would be nice to add godocs on these constants explaining difference between them and in which case one is shown vs the other.

kimlisa replied:
added comment

r0mant commented on lib/web/apiserver.go:
```diff
@@ -835,39 +840,27 @@ func (h *Handler) oidcLoginWeb(w http.ResponseWriter, r *http.Request, p httprou
 	if err != nil {
-		logger.WithError(err).Warn("Unable to extract CSRF token from cookie.")
-		return nil, trace.AccessDenied("access denied")
+		logger.Error(err)
```
`logger.WithError(err).Error("Failed to extract SSO parameters from request.")`

kimlisa replied:
revised

r0mant commented on lib/web/apiserver.go:
```diff
@@ -876,31 +869,26 @@ func (h *Handler) githubLoginWeb(w http.ResponseWriter, r *http.Request, p httpr
 	if err != nil {
-		logger.WithError(err).Warn("Unable to extract CSRF token from cookie.")
-		return nil, trace.AccessDenied("access denied")
+		logger.Error(err)
```
`logger.WithError(err).Error("Failed to extract SSO parameters from request.")`

kimlisa replied:
revised

r0mant commented on lib/web/apiserver.go:
```diff
@@ -1013,40 +1011,47 @@ func (h *Handler) oidcLoginConsole(w http.ResponseWriter, r *http.Request, p htt
+		if err := httplib.SafeRedirect(w, r, response.Req.ClientRedirectURL); err != nil {
+			logger.WithError(err).Error("Error parsing redirect URL.")
+			http.Redirect(w, r, loginFailedURL, http.StatusFound)
+		}
```
Nit: I know this is the last "if" block of the function, but I would add explicit `return nil, nil` inside the block anyway, in case someone adds more code below and forgets to add return here.

kimlisa replied:
revised

r0mant commented on lib/web/apiserver.go:
```diff
@@ -938,33 +926,40 @@ func (h *Handler) githubCallback(w http.ResponseWriter, r *http.Request, p httpr
-			return nil, trace.Wrap(err)
+
+		if err := SetSessionCookie(w, response.Username, response.Session.GetName()); err != nil {
+			logger.WithError(err).Error("Unable to set session cookie.")
```
So this is not a huge deal, but I'm wondering if you have considered making a some sort of wrapper method for these handlers which would do this logging/redirect instead of adding the same 2 lines of code to every "if err != nil" block. Would make the code a bit cleaner.

kimlisa replied:
revised

It made sense to create a redirect handler wrapper b/c we didn't use the return type values of the default handler. But I kept the logging of errors closer to origin b/c it was easier to customize

russjones reviewed (approved):
@kimlisa Approved once you make changes @r0mant suggested.

kimlisa committed:
Address CR 2
- Made redirect handler wrapper
- Moved redirect URL's const to lib/client/redirect.go
- Cleaned up error messages

kimlisa commented on lib/web/apiserver.go:
```diff
@@ -2328,3 +2336,35 @@ func makeTeleportClientConfig(ctx *SessionContext) (*client.Config, error) {
+
+	clientRedirectURL := query.Get("redirect_url")
+	if clientRedirectURL == "" {
+		return nil, trace.BadParameter("Missing redirect_url query parameter")
```
revised

r0mant commented on lib/client/redirect.go:
```diff
@@ -32,6 +32,18 @@ import (
+	// LoginSuccessRedirectURL is a redirect URL when login was successful without errors.
+	LoginSuccessRedirectURL = "/web/msg/info/login_success"
+
+	// LoginFailedRedirectURL is the default redirect URL when an sso error was encountered.
```
Nit: sso -> SSO

kimlisa replied:
revised

r0mant commented on lib/client/redirect.go:
```diff
@@ -32,6 +32,18 @@ import (
+	// LoginFailedRedirectURL is the default redirect URL when an sso error was encountered.
+	LoginFailedRedirectURL = "/web/msg/error/login"
+
+	// LoginFailedBadCallbackRedirectURL is a redirect URL when an sso error specific to
```
Nit: sso -> SSO

kimlisa replied:
revised

kimlisa committed:
Uppercase acronym
